File: Proxy/python/Proxy.py
import abc
from typing import Dict


class ISite(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def get_page(self, num: int) -> str:
        pass


class Site(ISite):

    def get_page(self, num: int) -> str:
        return 'page {}'.format(num)


class SiteProxy(ISite):

    def __init__(self, site: ISite):
        self.__site = site
        self.__cache: Dict[int, str] = {}

    def get_page(self, num: int) -> str:
        page: str = ''
        if self.__cache.get(num) is not None:
            page = self.__cache[num]
            page = 'from cache ' + page
        else:
            page = self.__site.get_page(num)
            self.__cache[num] = page
        return page


if __name__ == '__main__':
    my_site: ISite = SiteProxy(Site())

    print(my_site.get_page(1))
    print(my_site.get_page(2))
    print(my_site.get_page(3))

    print(my_site.get_page(2))

File: Proxy/c++/Proxy.cpp
#include <iostream>
#include <string>
#include <map>

class ISite
{
    public:
        virtual std::string GetPage(int num) = 0;
        virtual ~ISite() {}
};

class Site : public ISite
{
    public:
        std::string GetPage(int num) override;
};

class SiteProxy : public ISite
{
    private:
        ISite* site;
        std::map<int, std::string> cache;

    public:
        SiteProxy(ISite* site) : site(site) {}
        ~SiteProxy();
        std::string GetPage(int num) override;
};

int main()
{
    ISite* mySite = new SiteProxy(new Site());

    std::cout << mySite->GetPage(1) << std::endl;
    std::cout << mySite->GetPage(2) << std::endl;
    std::cout << mySite->GetPage(3) << std::endl;

    std::cout << mySite->GetPage(2) << std::endl;

    return 0;
}

std::string Site::GetPage(int num)
{
    std::string page = "page " + std::to_string(num);
    return page;
}

SiteProxy::~SiteProxy() noexcept
{
    delete site;
}

std::string SiteProxy::GetPage(int num)
{
    std::string page;
    if (cache.find(num) == cache.end())
    {
        page = site->GetPage(num);
        cache[num] = page;
    }
    else
    {
        page = cache[num];
        page.insert(0, "from cache ");
    }
    return page;
}